### PR TITLE
Fix #1538: The Predictoor page modal

### LINKED ISSUE #1538: [PdrDashboard, v0.5] The Predictoor Modal
Display specific Predictoor plots inside a modal
- When clicking on a specific row, the predictoor address should be saved and a component should open
- When clicking again the component should close
- Display following charts: Volume, Sales, Accuracy, Predictions, Stakes over time

### LINKED ISSUE #1546: Pdr Dashboard v0.5 - Add tests using dash duo

### PR DESCRIPTION
**Pull request by kdetry**
Closes #1538. Closes #1546.

### CODE REVIEW

**Comment by kdetry** [comment]
I will seperate the APY_APR issue from this PR

File: pdr_backend/pdr_dashboard/util/db.py [hunk 1]
```diff
@@ -248,29 +223,53 @@ def payouts(
...
+        if feed_addrs:
+            params.extend([f"'%{item}%'" for item in feed_addrs])
+        if predictoor_addrs:
+            params.extend([f"'%{item}%'" for item in predictoor_addrs])
+        if start_date:
+            params.append(str(start_date))
+
+        query = query % tuple(params)
+
+        print("query--->", query)
```
Comment (calina-c) [inline pdr_backend/pdr_dashboard/util/db.py:270]: please remove prints

**Review by calina-c — CHANGES_REQUESTED** [review]

File: pdr_backend/pdr_dashboard/pages/predictoors.py [hunk 2]
```diff
@@ -162,38 +163,36 @@ def get_data_for_predictoors_table(
...
         figures = FeedModalFigures()
         return [
-            dbc.ModalHeader("Loading feed data", id="feeds-modal-header"),
-            self.get_feed_graphs_modal_body(figures.get_figures()),
+            dbc.ModalHeader("Loading Predictoor data", id="predictoors-modal-header"),
+            self.get_predictoor_graphs_modal_body(figures.get_figures()),
         ]
 
-    def get_feed_graphs_modal_header(self, selected_row):
+    def get_predictoor_graphs_modal_header(self, selected_row):
```
[the remark this answers is not in the document]
Comment (kdetry) [reply pdr_backend/pdr_dashboard/pages/predictoors.py:179]: I can agree with you

File: pdr_backend/pdr_dashboard/callbacks/callbacks_feeds.py [hunk 3]
```diff
@@ -156,41 +161,31 @@ def clear_all_filters(btn_clicks):
         return ([], [], [], [], None, None, None, None, None, None, None, None, "")
 
     @app.callback(
-        Output("modal", "is_open"),
+        Output("feeds_modal", "is_open"),
```
Comment (calina-c) [inline pdr_backend/pdr_dashboard/callbacks/callbacks_feeds.py:164]: shouldn't this remain "modal"?

**Review by calina-c — COMMENTED** [review]
I thought we were going to reuse the same modal, but I see we still have different feed and predictoor modals.

I'm thinking simply add a common callback when selecting a row.
If it's selected from the feeds table, show the feeds results (charts).
If it's selected from the predictoors table, show the predictoors results (charts).

File: pdr_backend/pdr_dashboard/callbacks/callbacks_feeds.py [hunk 4]
```diff
@@ -207,8 +202,16 @@ def update_graphs(is_open, selected_rows, feeds_table_data):
         feed_figures: FeedModalFigures = get_feed_figures(payouts, subscriptions)
 
         children = [
-            feeds_page.get_feed_graphs_modal_header(selected_row),
-            feeds_page.get_feed_graphs_modal_body(feed_figures.get_figures()),
+            get_graphs_modal_header(
```
Comment (calina-c) [inline pdr_backend/pdr_dashboard/callbacks/callbacks_feeds.py:205]: ok, then to simplify, I think you can use just "feeds-modal"/"predictoors-modal" and selected row as the parameter to get_graphs_modal_header and get_graphs_modal_body. Assemble based on case, inside the function itself, to keep more code in helpers than in callbacks files.

**Review by calina-c — CHANGES_REQUESTED** [review]
added just a few minor nitpicks

**Comment by KatunaNorbert** [comment]
Feedback:

UI
- add gap between the charts so the yAxis legend is not right next to the previous chart
- display empty charts by default while the data is loading and then updated the charts so that by default the modal in not compressed and only showing the title. This was at one point the behaviour inside the feeds modal as well, I'm not sure what happened that now is changed

Functional
- Profit chart: display as in the home page showing the cumulative profit over time

**Review by KatunaNorbert — APPROVED** [review]

File: pdr_backend/pdr_dashboard/dash_components/modal.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            className="modal-graph-body",
+        )
+
+        return body
+
+    def create_figures(self):
+        selected_row = self.selected_row
+
+        if self.modal_id == "feeds_modal":
+            if selected_row:
```
Comment (calina-c) [inline pdr_backend/pdr_dashboard/dash_components/modal.py:77]: let's switch up the ifs here.
treat the empty selected_row case first
then move onto the rest

File: pdr_backend/pdr_dashboard/util/helpers.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import dash
+
+
+def toggle_modal_helper(
+    ctx,
+    selected_rows,
+    is_open_input,
+    safe_trigger_ids=None,
```
Comment (calina-c) [inline pdr_backend/pdr_dashboard/util/helpers.py:8]: I just remembered here that safe trigger ids completely depend on the id, I will adjust these.

File: pdr_backend/pdr_dashboard/dash_components/plots.py [hunk 7]
```diff
@@ -480,3 +567,48 @@ def get_feed_figures(payouts: Optional[List], subscriptions: List) -> FeedModalF
...
+
+    # Process each slot's payouts
+    for slot, payout_group in grouped_payouts.items():
+        slot_stake = sum(p["stake"] for p in payout_group)
+        slot_profit = sum(max(p["payout"], 0) - p["stake"] for p in payout_group)
+        slot_predictions = len(payout_group)
+
+        correct_predictions += sum(1 for p in payout_group if p["payout"] > 0)
+        total_predictions += slot_predictions
+        total_profit += slot_profit
```
Comment (kdetry) [inline pdr_backend/pdr_dashboard/dash_components/plots.py:602]: ty for this fix

File: pdr_backend/pdr_dashboard/dash_components/modal.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+def get_modal(
+    modal_id: str,
+):
+    return dbc.Modal(
+        ModalContent(modal_id).get_content(),
+        id=modal_id,
+    )
+
+
+class ModalContent:
```
Comment (kdetry) [inline pdr_backend/pdr_dashboard/dash_components/modal.py:23]: I like the class approach here

**Review by kdetry — COMMENTED** [review]
I have tested the code manually and reviewed it, and it looks good to me

**Review by calina-c — APPROVED** [review]

**Pull request merged**